### PR TITLE
[REEF-506] Make Driver call DriverRestartActiveContextHandler instead of have the Evaluator decide when to call it

### PR DESCRIPTION
**Pull request by afchung**
This addressed the issue by
  * Call restart handlers by checking the restart status based on DriverRestartManager instead of using the protobuf message from the recovered evaluator.

JIRA:
  [REEF-506](https://issues.apache.org/jira/browse/REEF-506)

### CODE REVIEW

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java [hunk 1]
```diff
@@ -45,30 +45,39 @@
 
   private boolean restartBegan;
   private boolean restartCompleted;
+  private boolean isRestartSet;
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java]: This set of mutable booleans smells like it might be better managed as a state machine. In any case, please document them as they all contain `restart` and are hard to distinguish.

Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java]: Also, the two new ones only seem to be used once, and could be replaced with a constant?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartUtilities.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * under the License.
+ */
+package org.apache.reef.driver.restart;
+
+import org.apache.reef.util.Optional;
+
+/**
+ * A static utilities class for simplifying calls to driver restart manager.
+ */
+public final class DriverRestartUtilities {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartUtilities.java]: This shouldn't have to be `public`, right?

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java [hunk 3]
```diff
@@ -45,30 +45,39 @@
...
+    this.isRestart = false;
+    this.isRestartSet = false;
     this.previousEvaluators = new HashSet<>();
     this.recoveredEvaluators = new HashSet<>();
   }
 
   @Override
-  public boolean isRestart() {
-    return driverRuntimeRestartManager.isRestart();
+  public synchronized boolean isRestart() {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java]: Also, these changes seem orthogonal to the intent of the PR?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java [hunk 4]
```diff
@@ -53,8 +56,22 @@
...
+  @Inject
+  private ContextRepresenters(final EvaluatorMessageDispatcher messageDispatcher,
+                              final ContextFactory contextFactory,
+                              final DriverRestartManager driverRestartManager) {
+    this(messageDispatcher, contextFactory, Optional.of(driverRestartManager));
+  }
+
+  private ContextRepresenters(final EvaluatorMessageDispatcher messageDispatcher,
+                              final ContextFactory contextFactory,
+                              final Optional<DriverRestartManager> driverRestartManager) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java]: This seems quite complicated. Have you considered a default implementation of `DriverRestartManager` that implements all its methods for the case that the Driver was *not*v restarted? That way, you can assume its presence in all other classes.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java [hunk 5]
```diff
@@ -210,7 +227,7 @@ private synchronized void onNewContext(final ReefServiceProtos.ContextStatusProt
         Optional.of(contextStatusProto.getParentId()) : Optional.<String>empty();
     final EvaluatorContext context = contextFactory.newContext(contextID, parentID);
     this.addContext(context);
-    if (contextStatusProto.getRecovery()) {
+    if (DriverRestartUtilities.isRestartAndIsPreviousEvaluator(driverRestartManager, context.getEvaluatorId())) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java]: wouldn't it be more natural to have that method in `DriverRestartManager`? E.g.

```
driverRestartManager.isRestartAndIsPreviousEvaluator(context.getEvaluatorId())
```

**Comment by markusweimer** [comment]
@afchung I've done a first pass.

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java (hunk 1 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java]: I see, state machine is probably the better pattern here.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java (hunk 5 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java]: I'd prefer to leave it as a static method, then it won't be up to the DriverRestartManager implementation to determine how to implement the method, since it's essentially just the concatenation of two method calls.

File: lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java (hunk 3 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManagerImpl.java]: It's just adding a layer of caching on top so it doesn't need to invoke the runtime restart manager every time.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java (hunk 4 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java]: I'll file a separate item for this. Doing this will change other portions of the code not relevant to the aim of this PR.

Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/context/ContextRepresenters.java]: Filed [REEF-587](https://issues.apache.org/jira/browse/REEF-587) for this.

**Comment by afchung** [comment]
@markusweimer Updated. Please have another look over. Thanks!

**Comment by afchung** [comment]
@markusweimer By the way, do you think we should remove the protobuf properties? Or should we leave them there for now?

**Comment by markusweimer** [comment]
> By the way, do you think we should remove the protobuf properties? Or should we leave them there for now?

Leave them for compatibility, but file a JIRA to remind us to delete them after 0.13 is released.

**Comment by markusweimer** [comment]
> @markusweimer Updated. Please have another look over. Thanks!

@afchung Thanks, I am doing another pass.

**Comment by markusweimer** [comment]
This looks good, I will test (and merge).
